crates/matrix-sdk/src/oidc/cross_process.rs

```diff
@@ -0,0 +1,322 @@
+        Ok(())
+    }
+}
+
+pub(super) struct CrossProcessRefreshLockGuard {
+    /// The hash for the latest session, either we one we knew, or the latest
+    /// one read from the database, if it was more up to date.
+    hash_guard: OwnedMutexGuard<Option<SessionHash>>,
+
+    /// Cross-process lock being hold.
+    _store_guard: CrossProcessStoreLockGuard,
+
+    /// Reference to the underlying store, for storing the hash of the latest
+    /// known session (as a custom value).
+    store: Store,
+
+    /// Do the in-memory hash and database hash mismatch?
+    ///
+    /// If so, this indicates that another process may have refreshed the token
+    /// in the background.
+    ///
+    /// We don't consider it a mismatch if there was no previous value in the
+    /// database. We do consider it a mismatch if there was no in-memory
+    /// value known, but one was known in the database.
+    pub hash_mismatch: bool,
+
+    /// Session hash previously stored in the DB. Used for debugging and testing
+    /// purposes.
+    db_hash: Option<SessionHash>,
+}
+
+impl CrossProcessRefreshLockGuard {
+    /// Updates the `SessionTokens` hash in-memory only.
+    fn save_in_memory(&mut self, hash: SessionHash) {
+        *self.hash_guard = Some(hash);
+    }
+
+    /// Updates the `SessionTokens` hash in the database only.
+    async fn save_in_database(
+        &self,
+        hash: SessionHash,
+    ) -> Result<(), CrossProcessRefreshLockError> {
+        self.store.set_custom_value(OIDC_SESSION_HASH_KEY, hash.0.to_le_bytes().to_vec()).await?;
+        Ok(())
+    }
+
+    /// Updates the `SessionTokens` hash in both memory and database.
+    ///
+    /// Must be called after a successful refresh.
+    pub async fn save_in_memory_and_db(
+        &mut self,
+        tokens: &SessionTokens,
+    ) -> Result<(), CrossProcessRefreshLockError> {
+        let hash = compute_session_hash(tokens);
+        // Note: fallible operation at the end, so the in-memory value at least is up to
+        // date.
+        self.save_in_memory(hash);
+        self.save_in_database(hash).await?;
+        Ok(())
+    }
+
+    /// Handle a mismatch by making sure values in the database and memory match
+    /// tokens we trust.
+    pub async fn handle_mismatch(
+        &mut self,
+        trusted_tokens: &SessionTokens,
+    ) -> Result<(), CrossProcessRefreshLockError> {
+        let new_hash = compute_session_hash(trusted_tokens);
+        trace!("Trusted OIDC tokens have hash {:?}; db had {:?}", new_hash, self.db_hash);
+
+        if let Some(db_hash) = &self.db_hash {
+            if new_hash != *db_hash {
+                // That should never happen, unless we got into an impossible situation!
+                // In this case, we assume the value returned by the callback is always
+                // correct, so override that in the database too.
+                tracing::error!("error: DB and trusted disagree. Overriding in DB.");
```

Comment:
Why not import the `error` macro? Also, please don't prefix an error-level tracing event with `error: ` ^^